### PR TITLE
Adds a peer-departed state

### PR DESCRIPTION
**Pull request by lazypower**
Related to https://github.com/juju-solutions/layer-etcd/issues/5

### CODE REVIEW

**Comment by lazypower** [comment]
I'm on the fence about this. This code in the interface works as is, but @johnsca says the -broken relation in reactive is broken. I'm only using the state as it gets fired on the single machine departing the cluster, which triggers self destruction, i mean, de-registration.

**Comment by mbruzek** [comment]
@johnsca What is the problem with the "-broken" hook that @chuckbutler talks about in the comment above?

**Comment by mbruzek** [comment]
I would think you would want to do this in the "-departed" hook, giving the code time to unregister with the cluster. My understanding of "-broken" is they relationship is completely severed. I am not against using "-broken" here, but this looks like a perfect example of "-departed" to me.

The relevant code reactive code that uses this state:  https://github.com/juju-solutions/layer-etcd/blob/master/reactive/etcd.py#L315

File: peers.py [hunk 1]
```diff
@@ -34,6 +34,13 @@ def peers_changed(self):
         if conv.get_remote('unit_name'):
             conv.set_state('{relation_name}.joining')
 
+    @hook('{peers:etcd}-relation-broken')
+    def peers_going_away(self):
+        conv = self.conversation()
```
Comment (mbruzek) [inline peers.py:24]: Couldn't you scope the conversation to UNIT right here and do this in '-departed'?

**Comment by mbruzek** [comment]
@chuckbutler pointed me to the conversation about this bug/problem: https://lists.ubuntu.com/archives/juju/2015-December/006245.html

I am +1 to this code if it works (as I am sure it does), but I am concerned as my understanding of the "-departed" is appropriate here. We would want each unit to run some code to remove the dying unit from their "peer list" configuration, as well as the dying unit to "unregister" itself from the master/leader or cluster.

**Comment by johnsca** [comment]
LGTM

File: peers.py [hunk 2]
```diff
@@ -1,69 +1,42 @@
...
-        @param: unit_name - charm name/unit# with the / stripped. eg: etcd1
+    def get_peers(self):
+        '''Return a list of peer names participating in the conversation scope
         '''
-        conversation = self.conversation(scope=scope)
-        conversation.set_remote(port=port, unit_name=unit_name,
-                                public_address=public_address)
-        conversation.remove_state('{relation_name}.joining')
-        conversation.remove_state('{relation_name}.declare_self')
+        peers = []
```
Comment (mbruzek) [inline peers.py:38]: There is a python one liner for these four lines. This way is fine, just flexing my python muscles.

**Comment by mbruzek** [comment]
+1 to this change, I like the removal of not_unless and making it easy to understand what is going on.

**Pull request merged**